### PR TITLE
[Maintenance] Fix removing Swiftmailer adapter

### LINKED ISSUE #167: Container exception if SwiftMailerBundle is installed but not loaded
**SyliusMailerBundle version affected**: 1.8.0

**Description**  
Upgrading from 1.7.0. to 1.8.0 causes the following error:

> In CheckExceptionOnInvalidReferenceBehaviorPass.php line 86: The service "sylius.email_sender.adapter.swiftmailer" has a dependency on a non-existent service "swiftmailer.mailer.default".

Version 1.8.0 decides, wether to remove or not remove the `sylius.email_sender.adapter.swiftmailer` service, by leveraging `ContainerBuilder::willBeAvailable`, which checks for the package to be installed, but not if the bundle is actually loaded.

We are using a custom sender adapter, which doesn't rely on SwiftMailer, thus the SwiftMailerBundle isn't loaded in the application. Although swiftmailer/swiftmailer and symfony/swiftmailer-bundle are installed as a transitive dependency of Sylius (1.11.7).

Hence `sylius.email_sender.adapter.swiftmailer` ends up in the container, while the dependency `swiftmailer.mailer.default` doesn't.

**Steps to reproduce**  
Upgrade to sylius/mailer-bundle 1.8.0 in a Sylius 1.11.7 application, without having the SwiftMailerBundle loaded (no entry in bundles.php).

**Possible Solution**  
Checking if the SwiftmailerBundle is actually loaded, would do the trick, although I'm not sure if it's possible at that point within the `SyliusMailerExtension` (maybe at least the order within bundles.php would matter?).

As a workaround we are adding `"conflict": { "sylius/mailer-bundle": "^1.8.0" }` to composer.json.

### PR DESCRIPTION
**Pull request by NoResponseMate**
| Q               | A
| --------------- | -----
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #167
| License         | MIT

### CODE REVIEW

**Review by coldic3 — APPROVED** [review]

**Comment by mbabker** [comment]
IMO this is the type of edge case that is better fixed in Symfony itself than trying to patch around it in downstream consumers of the framework.  `ContainerBuilder::willBeAvailable()` is by design based on the Composer manifest, this really should not need to be reliant on container parameters that can change at runtime.

Yes, it fixes the issue, but IMO it's only an issue because of the way a lot of dependencies are managed in the core `sylius/sylius` monopackage.

**Comment by NoResponseMate** [comment]
IMO there's a different problem and it's how the `::willBeAvailable()` is used.
I don't see the edge case here since the method does precisely what it states, checks whether the class is there and will be there without dev dependencies.
Now for whatever reason, it seemed to have been equated to `if the class is there, the service will be as well`, which is wrong.
The Swiftmailer case has been explained, but with Symfony Mailer it's similar, all you have to do is configure `framework.mailer.enabled` to `false` and you'll get the same exception.

A possible fix could be a CompilerPass, moving the adapter removing/resolving logic there, but removing based on whether the needed services are missing.

**Comment by mbabker** [comment]
A compiler pass is definitely cleaner for removing the adapter definitions if the dependencies are missing since by that point you'll be working with a fully built container (whereas when an extension is being loaded you can really only make checks based on the `kernel.bundles` parameter or use `ContainerBuilder::willBeAvailable()`, neither of which is 100% reliable here).

The pain point here is you then have to pass the user config through as temporary container parameters (makes me wish https://github.com/symfony/symfony/pull/47680 were finalized so that the convention could be adapted now) for both the renderer and sender adapters so that the `sylius.email_sender.adapter` and `sylius.email_renderer.adapter` aliases can still be set appropriately (based on user config first then falling back to feature detection), and the compiler pass should probably hard abort building the container if the user config is set to a service ID that doesn't exist (I know the bundle doesn't do that now, but it should to give a clearly actionable error instead of something more generic/cryptic from the framework).

(BTW, I did try a compiler pass-based approach at one point working on #102 but at this point, I don't remember why it didn't stick)

File: src/Bundle/DependencyInjection/SyliusMailerExtension.php [hunk 1]
```diff
@@ -44,7 +44,8 @@ protected function loadInternal(array $mergedConfig, ContainerBuilder $container
 
     private function configureSenderAdapter(array $mergedConfig, ContainerBuilder $container): void
     {
-        if (!ContainerBuilder::willBeAvailable('swiftmailer/swiftmailer', \Swift::class, ['symfony/swiftmailer-bundle'])) {
+        $bundles = $container->hasParameter('kernel.bundles') ? $container->getParameter('kernel.bundles') : [];
+        if (!is_array($bundles) || !isset($bundles['SwiftmailerBundle'])) {
```
Comment (Zales0123) [inline src/Bundle/DependencyInjection/SyliusMailerExtension.php:48]: I wonder, is there any way we could have a unified way to check it for both Swiftmailer and Symfony Mailer? On the other words: why does `willBeAvailable` work for SymfonyMailer but not for Swiftmailer? 🤔

**Review by Zales0123 — COMMENTED** [review]
I agree that it should be changed in the Symfony itself, but until it's done, we can always _fix_ it here. Worst case scenario is we release a patch version soon with this fix reverted 😄 🖖

File: src/Bundle/DependencyInjection/SyliusMailerExtension.php (hunk 1 above)
Comment (mbabker) [reply src/Bundle/DependencyInjection/SyliusMailerExtension.php:48]: It doesn't 100% work for either dependency at this point in the compile process.

Basing the Swift check on whether the bundle is installed will get a bit closer to being truthy than whether the library is available in the production environment.  Through the dependency chains, it'd be possible to have the `swiftmailer/swiftmailer` library and `symfony/swiftmailer-bundle` package installed, but if the bundle isn't activated in the kernel, then `ContainerBuilder::willBeAvailable()` would be true and the `kernel.bundles` parameter check would be false.

For the most reliable solution, you need to use a compiler pass that can accurately check if the dependent services of all of the adapters are available (Twig renderer, Swift sender, Symfony sender).  Twig and Swift both have the same kind of convention in that they rely on a separate library and bundle being installed and enabled, the Symfony mailer can be disabled through config on the FrameworkBundle.

**Comment by NoResponseMate** [comment]
I've made a [POC](https://github.com/NoResponseMate/SyliusMailerBundle/pull/1) using compiler passes and it seems to work just fine, but personally, I'm not too fond of the way aliases are handled.
Build parameters would solve this, but since both `1.8` and `2.0` support Symfony 5.4, it's out of the question.

**Comment by mbabker** [comment]
> Build parameters would solve this, but since both `1.8` and `2.0` support Symfony 5.4, it's out of the question.

I would've just used whatever convention the feature decided on and not worry about the behavioral difference between Symfony versions (so using the dot-prefixed parameter names as it proposes now, the param would always exist in the SF 5.4 container but on 6.3 it'd be removed).  But since that's not finalized in any way, what you've got without using parameters works well enough IMO.

**Comment by NoResponseMate** [comment]
I'm closing this since this fix has been replaced by #182

**Pull request closed without merging**